### PR TITLE
[SYCL] Duplicate address-taken functions in split device modules.

### PR DESCRIPTION
**Pull request by kbobrovs**
We conservatively assume that address-taken function can be used
in any split module, so it needs to be duplicated.

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

**Comment by kbobrovs** [comment]
@asudarsa, @v-klochkov, @bader - would appreciate your input on the patch.
sycl-post-link module splitting fetches only callgraphs starting from entry points.
Functions whose address are not considered entry points and might not be part
of static call graph, but still be used by functions in a split module.

**Comment by asudarsa** [comment]
Hi @kbobrovs 
Thanks for putting up the patch. I am not sure if these functions need to be entry points. I think they need to be included here: 'collectFunctionsToExtract'. Can you please comment?
Thanks again.

**Comment by kbobrovs** [comment]
@asudarsa, as we talked, we really never formally defined what are "entry points". Today they include SYCL_EXTERNAL functions (which are not kernels), to enable runtime linkage using explicit SYCL APIs. Address-taken functions seem similar - they need to be available for linkage.

**Comment by asudarsa** [comment]
Ok...That is agreeable. Thanks

**Comment by kbobrovs** [comment]
@asudarsa, further testing showed that treating address-taken functions as entry points crashes sycl-post-link when per-source split is requested, as such functions sometimes don't contain sycl-module-id attribute. It also occurred to me that if we ever want to link to address-taken function at runtime, it should really be declared as `SYCL_EXTERNAL`, in which case the function also becomes entry point. So I used your original proposal, just to add them to the `collectFunctionsToExtract`.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 1]
```diff
@@ -307,13 +315,25 @@ class CallGraph {
...
 
 void collectFunctionsToExtract(SetVector<const GlobalValue *> &GVs,
                                const EntryPointGroup &ModuleEntryPoints,
                                const CallGraph &Deps) {
   for (const auto *F : ModuleEntryPoints.Functions)
     GVs.insert(F);
+  // It is conservatively assumed that any address-taken function can be invoked
+  // or otherwise used by any function in any module split from the initial one.
+  // So such functions along with the call graphs they start are always
+  // extracted (and duplicated in each split module).
```
Comment (bader) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:332]: Can we move such function in a separate module and link only with modules, which use indirect invocation?

Comment (kbobrovs) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:332]: That could be done. But maybe as a separate (big) effort. There can potentially be other uses besides indirect invocation which would require availability of an address taken function (e.g. address comparison) and some data flow analysis might be needed. Conservative approach should be good enough to start with, I believe.

**Comment by asudarsa** [comment]
> > That's a good point. I think we need to define a new term for such function and avoid using "entry point" as this term already used in GPGPU world to denote device functions, which can be called from the host. I.e. functions we can "enter" from the host. If I understand it correctly, we can't call SYCL_EXTERNAL or address-taken functions from the host.
> 
> That's correct.
> 
> > Maybe "extern functions"?
> 
> I'm not sure. That could be confused with C's `extern` term. "exported functions" ? exported = entry points + SYCL_EXTERNAL

How about 'sycl_extern functions'. exported seems a bit vague in its meaning. Thanks

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp (hunk 1 above)
Comment (asudarsa) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:332]: > That could be done. But maybe as a separate (big) effort. There can potentially be other uses besides indirect invocation which would require availability of an address taken function (e.g. address comparison) and some data flow analysis might be needed. Conservative approach should be good enough to start with, I believe.

A TODO could be added for this. Thanks

Comment (kbobrovs) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:332]:
```suggestion
  // extracted (and duplicated in each split module).
  // TODO: try to determine which split modules really use address-taken
  // functions and only duplicate the functions in such modules. Note that usage
  // may include e.g. function address comparison w/o actual invocation.
```

Comment (kbobrovs) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:332]: Done

**Comment by kbobrovs** [comment]
> How about 'sycl_extern functions'. exported seems a bit vague in its meaning. Thanks

I think we are looking for a term in English, and `sycl_extern` seems like keyword :) We need a term to describe functions participating in the module symbol table used by the SYCL RT to
(1) lookup a kernel name to launch it
(2) lookup a function name to link to

**Comment by asudarsa** [comment]
> > How about 'sycl_extern functions'. exported seems a bit vague in its meaning. Thanks
> 
> I think we are looking for a term in English, and `sycl_extern` seems like keyword :) We need a term to describe functions participating in the module symbol table used by the SYCL RT to (1) lookup a kernel name to launch it (2) lookup a function name to link to

I am not able to think of anything better than 'exported'...

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks @kbobrovs

**Comment by kbobrovs** [comment]
The only failure is in CUDA LLVM Test Suite:
```
$ "env" "SYCL_DEVICE_FILTER=ext_oneapi_cuda:gpu,host" "SYCL_PI_CUDA_ENABLE_IMAGE_SUPPORT=1" "/__w/llvm/llvm/build/SYCL/Basic/image/Output/image_write_fp16.cpp.tmp.out"
note: command had no output on stdout or stderr
error: command failed with exit status: 255
```
The failure is unrelated. I don't have CUDA machine at hand, so I compared `sycl-post-link` output before and after my change - it is exactly the same (3 .bc files, the difference is only in clang version attribute ). So the patch can be merged, I believe.

**Pull request merged**